File: common/test/java/org/openqa/selenium/ElementEqualityTest.java
package org.openqa.selenium;

import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.openqa.selenium.Ignore.Driver.IE;
import static org.openqa.selenium.Ignore.Driver.IPHONE;
import static org.openqa.selenium.Ignore.Driver.REMOTE;
import static org.openqa.selenium.Ignore.Driver.SELENESE;

@Ignore(IPHONE)
public class ElementEqualityTest extends AbstractDriverTestCase {
  @Ignore(SELENESE)
  public void testElementEqualityShouldWork() {
    driver.get(pages.simpleTestPage);
    
    WebElement body = driver.findElement(By.tagName("body"));
    WebElement xbody = driver.findElement(By.xpath("//body"));

    assertEquals(body, xbody);
  }

  @Ignore(SELENESE)
  public void testElementInequalityShouldWork() {
    driver.get(pages.simpleTestPage);
    
    List<WebElement> ps = driver.findElements(By.tagName("p"));

    assertFalse(ps.get(0).equals(ps.get(1)));
  }

  @Ignore({IE, REMOTE, SELENESE})
  public void testFindElementHashCodeShouldMatchEquality() {
    driver.get(pages.simpleTestPage);
    WebElement body = driver.findElement(By.tagName("body"));
    WebElement xbody = driver.findElement(By.xpath("//body"));

    assertEquals(body.hashCode(), xbody.hashCode());
  }
  
  @Ignore({IE, REMOTE, SELENESE})
  public void testFindElementsHashCodeShouldMatchEquality() {
    driver.get(pages.simpleTestPage);
    List<WebElement> body = driver.findElements(By.tagName("body"));
    List<WebElement> xbody = driver.findElements(By.xpath("//body"));

    assertEquals(body.get(0).hashCode(), xbody.get(0).hashCode());
  }
}

File: common/test/csharp/webdriver-common-test/ElementEqualityTest.cs
using System;
using System.Collections.Generic;
using System.Text;
using NUnit.Framework;
using System.Collections.ObjectModel;

namespace OpenQA.Selenium
{
    [TestFixture]
    public class ElementEqualityTest : DriverTestFixture
    {
        [Test]
        public void ElementEqualityShouldWork()
        {
            driver.Url = (simpleTestPage);

            IWebElement body = driver.FindElement(By.TagName("body"));
            IWebElement xbody = driver.FindElement(By.XPath("//body"));

            Assert.AreEqual(body, xbody);
        }

        [Test]
        public void ElementInequalityShouldWork()
        {
            driver.Url = (simpleTestPage);

            ReadOnlyCollection<IWebElement> ps = driver.FindElements(By.TagName("p"));

            Assert.AreNotEqual(ps[0], ps[1]);
        }

        [Test]
        [IgnoreBrowser(Browser.IE, "HashCode valid only for elements with ID assigned by the driver. Not the case with IE.")]
        public void FindElementHashCodeShouldMatchEquality()
        {
            driver.Url = (simpleTestPage);
            IWebElement body = driver.FindElement(By.TagName("body"));
            IWebElement xbody = driver.FindElement(By.XPath("//body"));

            Assert.AreEqual(body.GetHashCode(), xbody.GetHashCode());
        }

        [Test]
        [IgnoreBrowser(Browser.IE, "HashCode valid only for elements with ID assigned by the driver. Not the case with IE.")]
        public void FindElementsHashCodeShouldMatchEquality()
        {
            driver.Url = (simpleTestPage);
            ReadOnlyCollection<IWebElement> body = driver.FindElements(By.TagName("body"));
            ReadOnlyCollection<IWebElement> xbody = driver.FindElements(By.XPath("//body"));

            Assert.AreEqual(body[0].GetHashCode(), xbody[0].GetHashCode());
        }
    }
}
